src/sources/timer/timerfd.rs

```diff
@@ -33,7 +36,19 @@ impl TimerScheduler {
 
     pub fn reschedule(&mut self, new_deadline: Instant) {
         let now = Instant::now();
-        let time = TimeSpec::from_duration(new_deadline.duration_since(now));
+
+        // We should handle the case when duration is zero. Since timerfd can't do that we pass the
+        // timer resolution, which is 1ns triggering the timer right away.
+        let time = TimeSpec::from_duration(
+            (now == new_deadline)
+                .then(|| TIMER_RESOLUTION)
+                .unwrap_or_else(|| {
+                    new_deadline
+                        .checked_duration_since(now)
+                        .unwrap_or(TIMER_RESOLUTION)
+                }),
+        );
```

Comment:
It's mostly about style, but what about this?

```suggestion
        let time = TimeSpec::from_duration(std::cmp::max(
            new_deadline.saturating_duration_since(now),
            TIMER_RESOLUTION
        ));
```

Reply:
I for whatever reason thought that duration isn't ord, but It seems like it is. Will update to your suggestion.